source/common/buffer/buffer_impl.h

```diff
@@ -93,16 +87,23 @@ class Slice {
    * @note Read-only implementations of Slice should return {nullptr, 0} from this method.
    */
   Reservation reserve(uint64_t size) {
-    if (reservation_outstanding_ || size == 0) {
+    if (size == 0) {
       return {nullptr, 0};
     }
+    if (data_ != 0 && data_ == reservable_) {
```

Comment:
Q: Why would this branch happen vs. being reset in the drain path? Should this be an assert instead?

Reply:
For a while, I couldn't decide whether to reset the offsets in the drain path or the reserve path, so it ended up in both. I'll replace this one with an assert.